Adding ZDC packer

matt2275 opened the pull request:
#### PR description:
Adding a ZDC packer to EventFilter/L1TRawtoDigi

Only output change should be L1TDigitoRaw now can include EtSumZDC info.

PR has no dependencies

#### PR validation:

PR tested with EventFilter/L1TRawtoDigi/test/test_RawToDigiToRawToDigi_GT_cfi.py where the output produces 2 sets of data:
1) Raw Dataunpacked to Digis
2) Raw Data unpacked to Digis then repacked and unpacked again

These 2 data sets were compared and were identical for EtSumZDC data.

I attached the summary of the RunTheMatrix test where there were some failures 
[RunTheMatrixSummary.txt](https://github.com/cms-sw/cmssw/files/14336808/RunTheMatrixSummary.txt)

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"0d31e0f9cfd80c2ac872d5dffcdbadf6e4a14384":{"files":["EventFilter/L1TRawToDigi/plugins/implementations_stage2/CommonTokens.h","EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.cc","EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCPacker.cc","EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCPacker.h"],"squashed":false,"time":1699368521},"11bb55646017da2acad57585c11a27a1cee5ddd4":{"files":["EventFilter/L1TRawToDigi/test/test_RawToDigiToRawToDigi_GT.py"],"squashed":false,"time":1708097921},"30adb21952727fbe8b72d2e6f11c004e26c75f3d":{"files":["EventFilter/L1TRawToDigi/python/gtStage2Raw_cfi.py"],"squashed":false,"time":1708370962},"8c156dd2e5b57c13c0c143da2c841a5aa6d27bc7":{"files":["EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.cc","EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTTokens.cc","EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCPacker.cc","EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCPacker.h","EventFilter/L1TRawToDigi/test/test_RawToDigiToRaw_GT.py"],"squashed":false,"time":1708097508},"9f8fad183c283a3fd50c7b7581a842c52bb9e463":{"files":[],"squashed":false,"time":1708354812},"c28a6f02b9d669d273629b0d33a8f50131c642e3":{"files":[],"squashed":false,"time":1708097578},"fc2afec96f88050053dcd6de9df157c1c1364853":{"files":[],"squashed":false,"time":1708370747}},"emoji":{},"last_seen_sha":"30adb21952727fbe8b72d2e6f11c004e26c75f3d","signatures":{}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44019/38948

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - EventFilter/L1TRawToDigi/test/test_RawToDigiToRaw_GT.py:
    - Added: 8c156dd2e5b57c13c0c143da2c841a5aa6d27bc7
    - Deleted: 11bb55646017da2acad57585c11a27a1cee5ddd4

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44019/38948/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44019/38948/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

missirol commented:
@matt2275 , please fix the code-format issues using `scram build code-format` as described above.

Just for my own understanding: does this PR as is take care of introducing all we need to re-emulate ZDC in the `L1REPACK` step ? Or, do we need any changes to files like the following ones ?
https://github.com/cms-sw/cmssw/blob/master/Configuration/StandardSequences/python/SimL1EmulatorRepack_Full_cff.py
https://github.com/cms-sw/cmssw/blob/master/Configuration/StandardSequences/python/SimL1EmulatorRepack_FullMC_cff.py
https://github.com/cms-sw/cmssw/blob/master/Configuration/StandardSequences/python/SimL1EmulatorRepack_FullSimTP_cff.py

FYI: @cms-sw/l1-l2 @cms-sw/hlt-l2

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44019/38965

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - EventFilter/L1TRawToDigi/test/test_RawToDigiToRaw_GT.py:
    - Added: 8c156dd2e5b57c13c0c143da2c841a5aa6d27bc7
    - Deleted: 11bb55646017da2acad57585c11a27a1cee5ddd4

cmsbuild commented:
A new Pull Request was created by  @matt2275 for master.

It involves the following packages:

- EventFilter/L1TRawToDigi (**l1**)


@cmsbuild, @aloeliger, @epalencia can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @dinyar, @thomreis, @missirol, @eyigitba this is something you requested to watch as well.
@rappoccio, @antoniovilela, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

matt2275 commented:
@missirol, I don't believe there is a L1 ZDC simulation yet so I don't think those files should be changed at the moment but I'm not an expert and could be mistaken.

aloeliger commented:
@missirol I don't think we should need to change anything much else in the configurations? This just adds the ZDC packer to the GT packing step, so anywhere that is used should get this by default. Most of the configurations you linked are defining L1 emulator unpack steps that otherwise, except for L1's emulation of L1, go un-used, or at least that is my interpretation. In any case, I don't see a GT step in any of those that _could_ be modified to account for ZDC I think.

aloeliger commented:
@matt2275 Did you have any plots or outputs from the RAW->DIGI->RAW test you performed to test the packer?

matt2275 commented:
![image](https://github.com/cms-sw/cmssw/assets/54369650/1823ef76-fc8d-45dd-9454-531ceadf96a5)
![image](https://github.com/cms-sw/cmssw/assets/54369650/35f236d3-c2ce-4f21-89a0-b9ee1ecca1ed)

Here is the ZDC energy distribution unpacked ( first image with title including gtstage2digis) and then ZDC energy unpacked packed and unpacked again ( second image with title including unpackgtStage2). Both distributions look identical and have identical statistics

aloeliger commented:
> ![image](https://private-user-images.githubusercontent.com/54369650/306685910-1823ef76-fc8d-45dd-9454-531ceadf96a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MzE5NjUsIm5iZiI6MTcwODUzMTY2NSwicGF0aCI6Ii81NDM2OTY1MC8zMDY2ODU5MTAtMTgyM2VmNzYtZmM4ZC00NWRkLTk0NTQtNTMxY2VhZGY5NmE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE2MDc0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5ZDI2MTM3ZDBhZDQwZjc3ZjMyZjM1OGI1YTViODcxNjRhMTJkNmFkMTRhYWRjMGExYzE2NDlkOTJiYmMxZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0-uLb-ogwMPT6cgtqiWAD25DvGd-iU7YOyr22p041bU) ![image](https://private-user-images.githubusercontent.com/54369650/306686034-35f236d3-c2ce-4f21-89a0-b9ee1ecca1ed.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1MzE5NjUsIm5iZiI6MTcwODUzMTY2NSwicGF0aCI6Ii81NDM2OTY1MC8zMDY2ODYwMzQtMzVmMjM2ZDMtYzJjZS00ZjIxLTg5YTAtYjllZTFlY2NhMWVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIxVDE2MDc0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY1MThmNDE2YzIzYjBjNGE2NWZlNjU0OWE2OWY2YmRjODhkNWQ4MDE4MzZhZmE5Y2Y5NGU3OGU2ODc2YWIzYWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MMKfn7RRkx5gm6fCwRgMm6ut7BkfVe383IsKovFqBso)
> 
> Here is the ZDC energy distribution unpacked ( first image with title including gtstage2digis) and then ZDC energy unpacked packed and unpacked again ( second image with title including unpackgtStage2). Both distributions look identical and have identical statistics

@eyigitba FYI

aloeliger commented:
please test

matt2275 commented:
I think the pull request is ready for review. In addition to making sure the repacked zdc sums match the original zdc sums. I made sure the new zdc packer wouldn't cause a crash if the collection name were incorrect or unused by the cms job. In that case, it gives a warning that the collection could not be found but doesn't throw any fatal errors. Is there any other things I should check before marking the pull request ready for review?

aloeliger commented:
@matt2275 I think that's good for now. Please mark this as good for review and maybe we can ping experts for any extra testing if needed.

aloeliger commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-541f95/37742/summary.html
**COMMIT**: 2b0b53da312cd2c639f71972d5ce29e7e3c963a9
**CMSSW**: CMSSW_14_1_X_2024-02-27-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/44019/37742/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 10 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-27-1100+541f95/61370/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 53 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-27-1100+541f95/61370/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3011576
 - DQMHistoTests: [Total failures: 6](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-27-1100+541f95/61370/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3011548
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-27-1100+541f95/61370/triggerResults): no differences found

aloeliger commented:
+l1

- I've spoken with the DPG and I think we are satisfied by the unpack-pack-unpack test done to test this packer.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @antoniovilela, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

antoniovilela commented:
@matt2275 @cms-sw/l1-l2 

It is difficult to follow the commit log. Could you squash the commits, even all into a single commit [1]?
Thanks.

[1]
- With “git cms-squash-topic” (https://cms-sw.github.io/tutorial-resolve-conflicts.html)
- Other refs.:
https://cms-sw.github.io/faq.html#how-do-i-collapse-multiple-commits-into-one
https://git-scm.com/book/en/Git-Tools-Rewriting-History

matt2275 commented:
Sorry for not seeing the message until sooner, it got buried in my email.  I'm looking at the instructions to git reset and the first commit of the PR and then amend all the following commits but when trying to push the branch, I run into the error message: Updates were rejected because a pushed branch tip is behind its remote. Should I force push to remote or something else?

antoniovilela commented:
> Sorry for not seeing the message until sooner, it got buried in my email.  I'm looking at the instructions to git reset and the first commit of the PR and then amend all the following commits but when trying to push the branch, I run into the error message: Updates were rejected because a pushed branch tip is behind its remote. Should I force push to remote or something else?

You need to force push because you are rewriting the history in the remote branch.
Please check first that the log is correct locally.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44019/39398

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44019/39398/cms-checkout-topic.log

matt2275 committed:
Adding ZDC Packer

Added ZDCPacker files: ZDCPacker.cc and ZDCPacker.h

Updated GTSetup.cc, GTTokens.cc and CommonTokens.h to work with ZDC Packer.

Updated gtStage2Raw_cfi.py to include ZDCPacker info.

Added test_RawToDigiToRawToDigi_GT.py to test the new ZDC Packer

matt2275 commented:
I just force pushed, but it caused some conflicts. I added all my changed to a new branch with a single commit here: https://github.com/matt2275/cmssw/tree/ZDCPackerUpdate/. I can close this PR and then submit a new PR with the ZDCPackerUpdate branch, Does that work?

antoniovilela commented:
hold

antoniovilela commented:
You don't need a new PR. If you have a branch with your previous commits, you can force push to the PR branch, see Recovering from a bad rebase in https://cms-sw.github.io/tutorial-resolve-conflicts.html

antoniovilela commented:
> You don't need a new PR. If you have a branch with your previous commits, you can force push to the PR branch, see Recovering from a bad rebase in https://cms-sw.github.io/tutorial-resolve-conflicts.html

And always have a backup branch.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-44019/39399

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTSetup.cc  modified in PR(s): #44222

cmsbuild commented:
Pull request has been put on hold by  @antoniovilela
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

cmsbuild commented:
Pull request #44019 was updated. @aloeliger, @epalencia, @cmsbuild can you please check and sign again.

antoniovilela commented:
unhold

antoniovilela commented:
please test

aloeliger commented:
+l1

re-sign

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @rappoccio, @sextonkennedy (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
